src/import/sst_service.rs

```diff
@@ -374,40 +430,101 @@ where
     }
 
     // Downloads KV file and performs key-rewrite then apply kv into this tikv store.
-    fn apply(&mut self, _ctx: RpcContext<'_>, req: ApplyRequest, sink: UnarySink<ApplyResponse>) {
+    fn apply(
+        &mut self,
+        _ctx: RpcContext<'_>,
+        mut req: ApplyRequest,
+        sink: UnarySink<ApplyResponse>,
+    ) {
         let label = "apply";
         let timer = Instant::now_coarse();
         let importer = Arc::clone(&self.importer);
-        let engine = self.engine.clone();
+        let router = self.router.clone();
         let limiter = self.limiter.clone();
         let start = Instant::now();
 
+        let mut futs = vec![];
+        let mut apply_resp = ApplyResponse::default();
+        let context = req.take_context();
+        let meta = req.get_meta();
+
+        match importer.do_download_kv_file(meta, req.get_storage_backend(), &limiter) {
+            Ok(temp_file) => {
+                let mut reqs = vec![];
+                let mut cmd_reqs = vec![];
+                let mut build_req_fn = self.build_apply_request(
+                    reqs.as_mut(),
+                    cmd_reqs.as_mut(),
+                    meta.get_is_delete(),
+                    meta.get_cf(),
+                    context.clone(),
+                );
+                match importer.do_apply_kv_file(
+                    meta.get_restore_ts(),
+                    temp_file,
+                    req.get_rewrite_rule(),
+                    &mut build_req_fn,
+                ) {
+                    Ok(range) => {
+                        drop(build_req_fn);
+                        if !reqs.is_empty() {
+                            let cmd = make_request(reqs.as_mut(), context);
+                            cmd_reqs.push(cmd);
+                        }
+                        for cmd in cmd_reqs {
+                            let (cb, future) = paired_future_callback();
+                            match router.send_command(
+                                cmd,
+                                Callback::write(cb),
+                                RaftCmdExtraOpts::default(),
+                            ) {
+                                Ok(_) => futs.push(future),
+                                Err(_e) => {
+                                    let mut import_err = kvproto::import_sstpb::Error::default();
```

Comment:
this is nested too deeply. can you do something like:

```rust
let result = (|| -> Result<()> {
    let temp_file = importer.do_download_kv_file(...)?;
    // etc.
    Ok(())
})();
if let Err(e) = result {
    apply_resp.set_error(e.into());
}
```

Reply:
Done